Allow when= to access value, parent, and root nodes

pivotaljohn opened the pull request:
Resolves #733

jtigger committed:
Allow when= to access value, parent, and root nodes

Resolves #733

jtigger committed:
Ref nodes by full path in violation messages

... because line numbers can sometimes be not present (e.g. data values
    being calculated) or shifted.

jtigger committed:
Reformat validation violation messages

... to improve readability.

Usability testing (#707) made clear that violation messages on a single
line are quite difficult to visually parse, making such error messages
less readable.

To support the reformatting, context is gathered at the time a violation
is detected. The message is formatted once all violations have
been collected.

Also includes minor improvements to "named" rule messages to improve
readability.

cppforlife commented on pkg/validations/validate.go:
```diff
@@ -163,30 +192,54 @@ func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.
+		for idx := 0; idx < whenFunc.NumParams(); idx++ {
+			name, _ := whenFunc.Param(idx)
+			switch name {
+			case "v", "val", "value":
+				kwargs = append(kwargs, starlark.Tuple{starlark.String(name), value})
+			case "p", "par", "parent":
+				kwargs = append(kwargs, starlark.Tuple{starlark.String(name), parent})
+			case "r", "root", "d", "doc", "document":
+				// When validation support is extended beyond Data Values and into documents in general, differentiate
+				//   between "root" and "document":
+				//   - since data values validation occurs on a document, "root" and "document" are synonyms, right now.
+				//   - once validations can be defined in templates, then "root" will be a DocSet.
+				kwargs = append(kwargs, starlark.Tuple{starlark.String(name), root})
```
ive been chewing on this a bit more today... care to do a quick zoom with me to talk it through?

pivotaljohn replied:
I would love to! I'll reach out.

pivotaljohn replied:
Summary:
- chief concern is that nowhere else in `ytt` is there an interaction like this: where the name of the argument of a supplied function has meaning outside of what the author wants.
- we landed on
  - one required positional parameter; it will receive the value being validated
  - a second optional positional parameter; when specified will be populated with a `struct` that has two attributes: `parent` and `root`.

cppforlife commented on pkg/yamltemplate/filetests/ytt-library/assert/max_len/value-fails-assertion.tpltest:
```diff
@@ -4,7 +4,7 @@ max_len_of_string: #@ assert.max_len(2).check("abc")
 
 ERR:
-- fail: length of 3 is more than 2
+- fail: length = 3
```
i feel like we lost too much of info of how to fix the problem (there might be a good reason for it though... lets talk).

also, i wonder if we can incorporate validation name here? (imagine if `assert.max_len(2)` was turned into a var then even with the source line it would be pretty hard to tell what validation is failing)

pivotaljohn replied:
Yeah, appreciate the attention to this. It's been challenging balancing the primary output (in the form of OOTB/named/keyword versions of the rules) from these more programmatic-based cases.

Definitely appreciate a space to talk through the trade-offs. 👍🏻

pivotaljohn replied:
Key concern here is that the output of the `check(v)` function, here, looks nothing like what one would expect when using this function in an assertion context outside of validations (e.g. in a unit test).

Of significant concern is making the experience of the end-user reading output from a failing validation at readable as possible... but not at the serious detriment of a package author's use of an assertion library for other uses.

Still chewing on this, but the going idea is to include _two_ functions: one that specifically works well with validations, another that is useful for an assertion result.

pivotaljohn replied:
Slept on it.

TL;DR. We'll defer. We have a good quality UX for our primary proto-persona (Consumers). The sub-par experience for our secondary group (Authors) points to a missing feature (which, when implemented ought to resolve the  design flaw) that can be done later without introducing a breaking change.

We deliberately deferred the feature of allowing the author of a rule to customize the "violation template" (i.e. the templated error message, described in the [proposal](https://github.com/vmware-tanzu/carvel/blob/004-schema-validation/proposals/ytt/004-schema-validation/README.md#specification) as the `message` argument). However, in our implementation, we failed to make room for this customization. This comes to a head as we attempt to come up with a violation message that serves _both_ the Configuration Consumer (who should see a message containing only the actuals when the rule fails since the overall message already contains the expected) and the Configuration Author (who would rightly expect the failure message to _include_ the expectation).

Given that:
- we want to release this feature as generally available this week;
- the gap in the DX in the Configuration Author case has a reasonable workaround (i.e. that they include the expectation in their test error messages, themselves);
- the feature that would resolve this tension is already defined;
- that this situation can be resolved without introducing a breaking change...

We'll accept this current implementation, as is; with the commitment to improve the DX for Configuration Authors, [soon](https://github.com/vmware-tanzu/carvel-ytt/issues/724).

jtigger committed:
Provide context nodes through an optional struct

...rather than mapping to a name argument; this is unlike any other
interaction in `ytt` authoring.
